docs/v1/P4Runtime-Spec.mdk

```diff
@@ -6304,6 +6328,409 @@ field [@ProtoAny]. At the moment, there is not any mechanism to extend the
 `p4.v1.TableEntry` message based on the value of architecture-specific table
 properties, but we may include on in future versions of the API.
 
+# GenericTable { #sec-generic-table}
+
+`GenericTable` is used to map non-PSA P4 externs or target-specific "fixed"
+features to their implementation in a generic way. The same can
+be achieved via `Extern`, but `GenericTable` provides a structured way in which
+every feature is represented as a set of match-fields and data-fields. The data
+consists of unions which are similar to actions. One use of `GenericTable` in a
+server backend is with TDI [Table Driven
+Interface](https://github.com/p4lang/tdi), but targets can use Generic table to
+map to their own specific feature implementations as well.
+
+## Generic Table ID Space { #sec-generic-table-id-space }
+
+A field in the `GenericTable` message in the P4info defines a
+`generic_table_type_id`.
+
+It is an 8-bit unsigned integer which uniquely identifies
+the generic_table_type in the context of the architecture.
+This value is in the range of `[0x00, 0xff]`.
+The ID in the preamble is created as `(0x18 << 24) + (generic_table_type_id
+<< 16) + (16-bit ID generated by the P4 compiler)`. The `generic_table_type`
+depicts a new table "type" which would have similar properties.
+
+Let's say we have 2 target specific P4 externs called `XyzCounter` and
+`XyzMeter` and each of them has 2 instances. `XyzCounter` has
+`generic_table_type` as 1 and `XyzMeter` as 2.
+
+```text
+XyzCounter1 : 0x18 01 00 01
+XyzCounter2 : 0x18 01 00 02
+
+XyzMeter1 : 0x18 02 00 01
+XyzMeter2 : 0x18 02 00 02
+```
+The above system ensures that the instance space allows for 65355
+instances. The table type space is limited to 255.
+
+Some optional guidelines:
+* For P4 externs, any new extern to have a new table type.
+* For target-specific fixed function blocks, options are
+  * All fixed tables grouped under one giant non_p4 table type
+  * Tables divided on the basis of categories. Categories are defined in the
+    spec but aren't part of the P4info.
+  * Every table has its own table-type. This can however lead to running out
+    of the 8 bit type ID space soon.
+Note that this value does not need
+to be unique across all architectures from all organizations, since at any
+given time every device managed by a P4Runtime server maps to a single P4Info
+message and a single architecture.
+
+
+## GenericTable P4info { #sec-p4info-generic-table}
+
+`GenericTable` message in P4info defines a Generic Table. It contains several
+instances in `GenericTableInstance` rather than duplicating it every time for a
+new instance just like in `Extern`.
+It defines the following fields
```

Comment:
The wording of the paragraph above confuses me.  Something like the following would be clearer to me (if it is correct):

A `GenericTable` message in P4info defines one type of generic table, but not a specific instance of a generic table.  Each instance of a generic table is defined by a `GenericTableInstance` message.  This separation allows field values that are common for all instances with the same generic table type to be defined in one place, the `GenericTable` message.  A `GenericTable` message contains the following fields: